### ISSUE TITLE
Should getQty() return int/float or string?

### ISSUE THREAD

**Issue by sreichel** [magento/magento2#18094]
### Summary

- Magento 2.2-dev/2.3-dev
- PHP7.1 (strict)

PHPDocs blocks say `getQty()` should return float/int, but in most cases it returns a string. This is no problem until I use strict type declaration for my own extensions and trust in annotations.

### Examples

Magento\Catalog\Model\Product

```
    /**
     * Get quantity for product
     *
     * @return float
     */
    public function getQty()
    {
        return $this->getData('qty');
    }
```
I'd expect to get a float returned, but you'll get a string ...

### Proposed solution

1. how to fix?
a. cast to float in getters?
b. cast to float in setters?

2. where to fix?
a. fix step by step?
b. fix all occurrence?

Please let me know, how to fix before i create PRs.

**Comment by magento-engcom-team** [comment]
Hi @sreichel. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@sreichel do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by ghost** [comment]
HI @sreichel  How did you use the getQty() ? i'm using the  `Magento\CatalogInventory\Api\StockStateInterface;`
method  `$this->stockInvertory->getStockQty();`
and was have correct return type 'float'
![selection_104](https://user-images.githubusercontent.com/40756439/45629925-3b772d80-baa0-11e8-916a-b9802d82c830.png)

**Comment by sreichel** [comment]
Hi @engcom-backlog-nazar,

I observe `checkout_cart_product_add_after` event to manipulate cart items.

Simplified code:

```
$product = $observer->getProduct();
$qty = $product->getQty();
$result = $this->doSomething($qty);
```

In this case `Magento\Catalog\Model\Product::getQty()` is used.

**Comment by ghost** [comment]
@sreichel But i'm have correct return type.
![selection_105](https://user-images.githubusercontent.com/40756439/45632664-88f69900-baa6-11e8-9857-4d3d9f43200e.png)

**Comment by ghost** [comment]
@sreichel Ok i'm reproduce this with another number of qty like 110 :)
![selection_106](https://user-images.githubusercontent.com/40756439/45633148-b859d580-baa7-11e8-8578-5de2646cd986.png)

**Comment by sreichel** [comment]
@engcom-backlog-nazar very strange. Again tested on 2.2-dev ...

![m2-18094](https://user-images.githubusercontent.com/5022236/45633095-39609f00-ba9f-11e8-9d62-c8c5d04564b5.PNG)

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95061`, `MAGETWO-95062` were created

**Comment by ghost** [comment]
@sreichel, thank you for your report.
We've acknowledge the issue and added to our backlog.

**Comment by jayankaghosh** [comment]
I am working on this

**Comment by slavvka** [comment]
Hi @sreichel. Thank you for your report. 
The issue has been fixed in magento/magento2#18149 by @jayankaghosh in 2.3-develop branch
Related commit(s):
 - [76676ce5d0648290538ba7e1d8ac12b6d0548091](https://github.com/magento/magento2/commit/76676ce5d0648290538ba7e1d8ac12b6d0548091)
 - [2ad1031531394b0470dfe12f25a557dc84d1ac21](https://github.com/magento/magento2/commit/2ad1031531394b0470dfe12f25a557dc84d1ac21)

The fix will be available with the upcoming 2.3.1 release.

**Comment by okorshenko** [comment]
Hi @sreichel. Thank you for your report. 
The issue has been fixed in magento/magento2#18424 by @lewisvoncken in 2.2-develop branch
Related commit(s):
 - [3fc107ad5426ccf73566d562cf643c9e94e27d5b](https://github.com/magento/magento2/commit/3fc107ad5426ccf73566d562cf643c9e94e27d5b)
 - [fa529366822a4e049ddcab820212af24ee931c2b](https://github.com/magento/magento2/commit/fa529366822a4e049ddcab820212af24ee931c2b)

The fix will be available with the upcoming 2.2.8 release.
